x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java

```diff
@@ -121,14 +128,41 @@ public ScoreMode scoreMode() {
 
     @Override
     public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        throw new UnsupportedOperationException("Shouldn't be here");
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub, AggregationExecutionContext aggCtx)
+        throws IOException {
         assert sub.isNoop() : "Expected noop but was " + sub.toString();
 
         BucketedSort.Leaf leafSort = sort.forLeaf(ctx);
 
         return new LeafBucketCollector() {
+
+            CompetitiveIterator competitiveIterator;
+
+            @Override
+            public DocIdSetIterator competitiveIterator() throws IOException {
+                if (aggCtx != null && aggCtx.getTsid() != null) {
+                    // TODO: we need to check that the sorting is that in @timestamp in descending order
+                    SortedDocValues tsids = DocValues.getSorted(ctx.reader(), TimeSeriesIdFieldMapper.NAME);
+                    competitiveIterator = new CompetitiveIterator(tsids);
+                }
+                intArray = bigArrays().newIntArray(1);
```

Comment:
I don't feel great about having side effects in `competitiveIterator` or needing to change the `collect` logic. What about only optimizing the case when `size` is 1, which I would expect to be by far the main case we're interested in optimizing?

Reply:
ok, we can revisit this when there is a clear use case.

Reply:
I changed the implementation so we are using now the introduced DocsPerOrdIterator which does not introduce side effects in competitiveIterator. Still we need to modify the collect logic as we need to tell the competitive iterator that a doc has been consumed. We don't know which document might intersect with a possible query so we need this information from collect.